src/main/java/io/pinecone/PineconeDataPlaneClient.java

```diff
@@ -0,0 +1,222 @@
+        UpdateRequest.Builder updateRequest = UpdateRequest.newBuilder();
+
+        if (id == null) {
+            throw new PineconeValidationException("Invalid Update Request. Vector ID must be present");
+        }
+        updateRequest.setId(id);
+
+        if (values != null) {
+            updateRequest.addAllValues(values);
+        }
+
+        if (metadata != null) {
+            updateRequest.setSetMetadata(metadata);
+        }
+
+        if (namespace != null && !namespace.isEmpty()) {
+            updateRequest.setNamespace(namespace);
+        }
+
+        if (sparseIndices != null && sparseValues != null) {
+            if (sparseIndices.size() != sparseValues.size()) {
+                throw new PineconeValidationException("Invalid Update Request. Please ensure that both sparse indices and values are of the same length.");
+            }
+
+            updateRequest.setSparseValues(SparseValues.newBuilder()
+                    .addAllIndices(convertUnsigned32IntToSigned32Int(sparseIndices))
+                    .addAllValues(sparseValues)
+                    .build());
+        }
+
+        return blockingStub.update(updateRequest.build());
+    }
+
+    public DeleteResponse delete(List<String> ids,
+                                 boolean deleteAll,
+                                 String namespace, Struct filter) {
+        DeleteRequest.Builder deleteRequestBuilder = DeleteRequest.newBuilder().setDeleteAll(deleteAll);
+
+        if ((ids != null && !ids.isEmpty()) && filter != null) {
+            throw new PineconeValidationException("Invalid Delete Request. Both IDs and filter cannot be present.");
+        }
+
+        if (ids != null && !ids.isEmpty()) {
+            deleteRequestBuilder.addAllIds(ids);
+        }
+
+        if (namespace != null && !namespace.isEmpty()) {
+            deleteRequestBuilder.setNamespace(namespace);
+        }
+
+        if (filter != null) {
+            deleteRequestBuilder.setFilter(filter);
+        }
+
+        return blockingStub.delete(deleteRequestBuilder.build());
+    }
+
+    public DescribeIndexStatsResponse describeIndexStats(Struct filter) {
+        DescribeIndexStatsRequest.Builder describeIndexStatsRequest = DescribeIndexStatsRequest.newBuilder();
+
+        if (filter != null) {
+            describeIndexStatsRequest.setFilter(filter);
+        }
+
+        return blockingStub.describeIndexStats(describeIndexStatsRequest.build());
+    }
+
+    // ToDo: Add tests
+    private Iterable<? extends Integer> convertUnsigned32IntToSigned32Int(Iterable<? extends Long> unsigned32IntValues) {
+        List<Integer> int32Values = new ArrayList<>();
+        for (Long value : unsigned32IntValues) {
+            if (value < 0 || value > 0xFFFFFFFFL) {
+                throw new PineconeValidationException("Sparse indices are out of range for unsigned 32-bit integers.");
+            }
+            int32Values.add(value.intValue());
+        }
+        return int32Values;
+    }
```

Comment:
Even though it's private, this feels out of place here since this class is mainly about defining the public interface for our data plane functionality. I would make it a static method on some kind of small utility class with thorough unit testing. Also some comments to explain the purpose of this will be useful in the future after we forget why this was necessary.

Reply:
Yes, I'll be moving it out of this class because of the reasons you mentioned and another reason is if I want to test this function, I'll need to instantiate blockingStub.